src/vmm/benches/memory_access.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+use vm_memory::{GuestAddress, GuestMemory};
+use vmm::resources::VmResources;
+use vmm::vmm_config::machine_config::{HugePageConfig, VmConfig};
+
+fn bench_single_page_fault(c: &mut Criterion, configuration: VmResources) {
+    c.bench_function("page_fault", |b| {
+        b.iter_batched(
+            || {
+                let memory = configuration.allocate_guest_memory().unwrap();
+                let ptr = memory
+                    .get_slice(GuestAddress(0), 1)
+                    .unwrap()
+                    .ptr_guard_mut()
+                    .as_ptr();
+
+                // fine to return both here, because ptr is not a reference into `memory` (e.g. no
+                // self-referential structs are happening here)
+                (memory, ptr)
+            },
+            |(_, ptr)| unsafe {
+                // Cause a single page fault
+                ptr.write_volatile(1);
+            },
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn bench_4k_page_fault(c: &mut Criterion) {
+    bench_single_page_fault(
+        c,
+        VmResources {
+            vm_config: VmConfig {
+                vcpu_count: 1,
+                mem_size_mib: 2,
+                ..Default::default()
+            },
+            ..Default::default()
+        },
+    )
+}
+
+pub fn bench_2m_page_fault(c: &mut Criterion) {
+    bench_single_page_fault(
+        c,
+        VmResources {
+            vm_config: VmConfig {
+                vcpu_count: 1,
+                mem_size_mib: 2,
+                huge_pages: HugePageConfig::Hugetlbfs2M,
+                ..Default::default()
+            },
+            ..Default::default()
+        },
+    )
+}
+
+criterion_group! {
+    name = memory_access_benches;
+    config = Criterion::default().noise_threshold(0.05);
+    targets = bench_4k_page_fault, bench_2m_page_fault
```

Comment:
is this also testing the case with memfd on/off? it not, should we add also that to the benchmarks?

Reply:
The idea Nikita and me had was that the benchmark should be oblivious to the actual memory allocation mechanism being used, and notify us if a change to the memory allocation mechanism (e.g. the change from MAP_ANONYMOUS to memfd_create in 1.6, or a potential future change to guest_memfd) incurs a higher page fault overhead.

Reply:
but we're only benchmarking one of the two branches of the function, while it would cost nothing to benchmark also the other part.
what if we change the backing in vhost-enabled vms?

Reply:
If we really wanted to, then we could change the function to accept a parameter for how the memory has to be allocated, so that we can easily test both paths.
But I wouldn't delay merging this PR on that tbh